### ISSUE TITLE
await HubConnection.Stop();

### ISSUE THREAD

**Issue by NVentimiglia** [SignalR/SignalR#645]
HubConnection.Stop() is a long lasting method. It seems to take longer then Connect.

Shouldn't this method be async ?

After some more testing, Ive also noticed that Stop() takes a long time to run. Like 40 seconds. Making my users wait 40 seconds while the disconnect magic happens seems wrong.

I am using a WinRT / Metro client.
@gissues:{"order":25,"status":"backlog"}

**Comment by sisterray** [comment]
HI,
Is this still an issue? I am using the SignalR PCL library version 2.1.2 (mainly because I could not get the 2.2.0 nuget package to install). Is there fix for this coming along in v 2.2.1.

I am seeing the HubConnection.Stop() method hanging (and never returning) even if there is no messages being sent or received on that connection. This is even when I specify a timeout. However I am making web api requests through other httpclient connections on other threads. This is in an iOS app.

Is there anyway to stop this deadlock when calling stop. This is a major issue for me at the moment, and I am hoping someone can help.

**Comment by halter73** [comment]
@sisterray Do you have a stack trace showing where HubConnection.Stop() is hanging?

**Comment by sisterray** [comment]
All I am doing is calling Stop on my HubConnection. I am using a timeout of 3 seconds. I I attempt to do this with other web api requests in progress (different threads using separate http clients) it hangs. If I do not have other web api requests in progress it times out correctly

**Comment by sisterray** [comment]
Do you think this change [Remove _stateLock from Connection.Stop to prevent deadlocks during Abort] (https://github.com/SignalR/SignalR/pull/2371) would have fixed this in 2.2.0? If so can you tell me when 2.2.1 is likely to be released. I think 2.2.1 fixes the issue which prevents me installing it using nuget into my Xamarin project

**Comment by sisterray** [comment]
I am finding that this is not always happening. I am assuming it the deadlock issue described in this thread.

**Comment by halter73** [comment]
@sisterray It seems pretty likely that this is fixed in 2.2.0. Unfortunately, I can't give you an exact date on the 2.2.1 release other than to say it's definitely not going to happen before next Thursday.

Have you tried changing your portable profile? http://stackoverflow.com/a/32123476/719967

**Comment by sisterray** [comment]
Thanks @halter73 I tried getting this working a while ago. I'm using mvvmcross and the profiles seem to clash. I'll maybe try to download the 2.2.0 nugget package locally and temporary use the library directly to see if it fixes it.

**Comment by sisterray** [comment]
l

**Comment by munir1690** [comment]
Hey I am facing same problem. Any way around this issue?

**Comment by sisterray** [comment]
You could try a couple of things
1. Upgrade to 2.2.1
2. Ensure that you cleanly manage the connection, e.g. Ensure the the previous connection is cleanly disposed of before setting your connection to a new one.
3. Don't abort or dispose of a connection that failed to connect, e.g. One with no connection Id. I found this seemed to work on iOS but crashed on Android. I suspect it didn't really work on iOS but eventually causes the locking issue.

Since I have done this the issue has gone.

**Comment by apathytomorrow** [comment]
I have this issue aswell. Sometimes connection.Stop() executes in a few seconds, other times it can take up to 30.

**Comment by moozzyk** [comment]
@apathytomorrow - you might be hitting a different issue. Without details it is not possible to tell what's happening.

**Comment by patsadow2** [comment]
2.4.0 Still the same issue long hubConnection.Stop(); of application in Xamarin.Froms ;/

**Comment by kunal2387** [comment]
SignalR version 2.2.3
In case of Timeout after 50 seconds (Keeping print preview window open for more than 50 seconds in IE 11 and then closing it causes timeout exception), 
SignalR calls the stop method without notifying the server.
connection.stop(/* async */ false, /* notifyServer */ false);
Because of which server Hub method, 
public override Task OnDisconnected(bool stopCalled)
is not called immediately.

And after this the client disconnect method is called,
$.connection.hub.disconnected(function () {
   //start new connection
  $.connection.hub.start({ transport: 'longPolling', xdomain: true }, function () {});
});

The  problem is that I am maintaining list of user depending upon session and connecting Ids on server when hub server method 
public override Task OnConnected()
is called.
And I do remove it on hub server method: 
public override Task OnDisconnected(bool stopCalled)
is called.

And in case of timeout, the hub server method: public override Task OnDisconnected(bool stopCalled)
is not called immediately.
And from the client disconnected method I am starting a new signalR connection,
As the user is already present in list I am facing , user duplication issue.

In short,
When timeout occurs, after Hub Server method : public override Task OnDisconnected(bool stopCalled) is called, then only I need to raise a new start request from client side.
How can I achieve this,

What I am doing currently is calling the start after 35 seconds as signalR connection timeout is 30seconds by default.

$.connection.hub.disconnected(function () {
                
		window.setTimeout(function () {
			$.connection.hub.start({ transport: 'longPolling', xdomain: true }, function () {});
		}, 35000);
	
	return true;
});

Any other way I can achieve this instead if waiting for 35 seconds.
